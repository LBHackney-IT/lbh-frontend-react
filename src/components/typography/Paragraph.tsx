import classNames from "classnames";
import { nullValuesAsUndefined } from "null-as-undefined";
import PropTypes from "prop-types";
import React from "react";

import { Attributes, DataAttributes } from "../../helpers/Attributes";

import "lbh-frontend/lbh/core/_typography.scss";

/**
 * The possible sizes for {@link Paragraph}.
 */
export enum ParagraphSize {
  XS = "xs",
  S = "s",
  M = "m",
  L = "l",

  /**
   * An alias for styling paragraph text as a lead paragraph.
   */
  Lead = "lead",
}

/**
 * The proptypes for {@link Paragraph}.
 *
 * This also supports all `aria-*` and `data-*` attributes.
 *
 * @noInheritDoc
 */
export interface ParagraphProps extends React.AriaAttributes, DataAttributes {
  id?: string | null;
  className?: string | null;
  /**
   * The size of the paragraph text.
   *
   * Defaults to standard sized paragraph text.
   */
  size?: ParagraphSize | null;
  /**
   * The Paragraph text.
   */
  children: React.ReactNode;
}

/**
 * A component for creating paragraph text.
 */
export const Paragraph: React.FunctionComponent<ParagraphProps> = (
  props: ParagraphProps
): React.ReactElement => {
  const { id, className, size, children } = nullValuesAsUndefined(props);

  const extraAttributes = Attributes.ariaAndData(props);

  return (
    <p
      id={id}
      className={classNames(`lbh-body${size ? `-${size}` : ""}`, className)}
      {...extraAttributes}
    >
      <div>{children}</div>
    </p>
  );
};

Paragraph.propTypes = {
  id: PropTypes.string,
  className: PropTypes.string,
  size: PropTypes.oneOf<ParagraphSize>([
    ParagraphSize.XS,
    ParagraphSize.S,
    ParagraphSize.M,
    ParagraphSize.L,
  ]),
  children: PropTypes.node.isRequired,
} as PropTypes.ValidationMap<ParagraphProps>;
