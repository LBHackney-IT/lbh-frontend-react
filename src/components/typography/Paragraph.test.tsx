import React from "react";
import { create } from "react-test-renderer";

import { Paragraph, ParagraphSize } from "./Paragraph";

it("renders correctly with all props", () => {
  const component = create(
    <Paragraph id="test" className="testClass" size={ParagraphSize.L}>
      Test
    </Paragraph>
  );

  expect(component).toMatchInlineSnapshot(`
    <p
      className="lbh-body-l testClass"
      id="test"
    >
      Test
    </p>
  `);
});

it("renders correctly without optional props", () => {
  const component = create(<Paragraph>Test</Paragraph>);

  expect(component).toMatchInlineSnapshot(`
    <p
      className="lbh-body"
    >
      Test
    </p>
  `);
});
