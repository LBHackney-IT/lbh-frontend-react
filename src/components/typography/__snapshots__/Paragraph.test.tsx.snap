// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with all props 1`] = `
<p
  className="lbh-body-l testClass"
  id="test"
>
  Test
</p>
`;

exports[`renders correctly without optional props 1`] = `
<p
  className="lbh-body"
>
  Test
</p>
`;
